### PR TITLE
Fix layering of Implement-Interface code-fixer vs language service

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/60990

### CODE REVIEW

File: src/Features/CSharp/Portable/ImplementInterface/CSharpImplementInterfaceCodeFixProvider.cs [hunk 1]
```diff
@@ -2,67 +2,28 @@
...
 namespace Microsoft.CodeAnalysis.CSharp.ImplementInterface;
 
 [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.ImplementInterface), Shared]
 [ExtensionOrder(After = PredefinedCodeFixProviderNames.ImplementAbstractClass)]
-internal class CSharpImplementInterfaceCodeFixProvider : CodeFixProvider
+[method: ImportingConstructor]
+[method: SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
+internal sealed class CSharpImplementInterfaceCodeFixProvider() : AbstractImplementInterfaceCodeFixProvider<TypeSyntax>
 {
-    private readonly Func<TypeSyntax, bool> _interfaceName = n => n.Parent is BaseTypeSyntax && n.Parent.Parent is BaseListSyntax && ((BaseTypeSyntax)n.Parent).Type == n;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ImplementInterface/CSharpImplementInterfaceCodeFixProvider.cs:23]: became impl of abstract method instead below.

File: src/Features/CSharp/Portable/ImplementInterface/CSharpImplementInterfaceCodeFixProvider.cs [hunk 2]
```diff
@@ -2,67 +2,28 @@
...
-    [ImportingConstructor]
-    [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
-    public CSharpImplementInterfaceCodeFixProvider()
-    {
-    }
-
     public sealed override ImmutableArray<string> FixableDiagnosticIds { get; }
         = [CS0535, CS0737, CS0738];
 
-    public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ImplementInterface/CSharpImplementInterfaceCodeFixProvider.cs:38]: core logic moved intto new abstract base class to share between C# and VB.

File: src/Features/CSharp/Portable/ImplementInterface/CSharpImplementInterfaceCodeFixProvider.cs [hunk 3]
```diff
@@ -2,67 +2,28 @@
...
-        if (actions.IsDefaultOrEmpty)
-            return;
-
-        context.RegisterFixes(actions, context.Diagnostics);
-    }
-
-    public sealed override FixAllProvider GetFixAllProvider()
-        => WellKnownFixAllProviders.BatchFixer;
+    protected override bool IsTypeInInterfaceBaseList(TypeSyntax type)
+        => type.Parent is BaseTypeSyntax { Parent: BaseListSyntax } baseTypeParent && baseTypeParent.Type == type;
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/ImplementInterface/CSharpImplementInterfaceCodeFixProvider.cs:28]: this is that lambda from above.

File: src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+                }
+
+                break;
+            }
+        }
+    }
+
+    private static string GetTitle(ImplementInterfaceOptions options)
+    {
+        if (options.ImplementDisposePattern)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs:76]: computation of the code action title moved here.

File: src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+                    using var _ = ArrayBuilder<CodeAction>.GetInstance(out var codeActions);
+                    await foreach (var implementOptions in GetImplementOptionsAsync(document, info, cancellationToken))
+                    {
+                        var title = GetTitle(implementOptions);
+                        var equivalenceKey = GetEquivalenceKey(info, implementOptions);
+                        codeActions.Add(CodeAction.Create(
+                            title,
+                            cancellationToken => service.ImplementInterfaceAsync(
+                                document, info, implementOptions, generationOptions, cancellationToken),
+                            equivalenceKey));
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs:62]: we now create a normal codeaction.  no special subclasses.

File: src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        ImplementInterfaceOptions options)
+    {
+        var interfaceType = state.InterfaceTypes.First();
+        var typeName = interfaceType.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat);
+        var assemblyName = interfaceType.ContainingAssembly.Name;
+
+        // Legacy part of the equivalence key.  Kept the same to avoid test churn.
+        var codeActionTypeName = options.ImplementDisposePattern
+            ? "Microsoft.CodeAnalysis.ImplementInterface.AbstractImplementInterfaceService+ImplementInterfaceWithDisposePatternCodeAction"
+            : "Microsoft.CodeAnalysis.ImplementInterface.AbstractImplementInterfaceService+ImplementInterfaceCodeAction";
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs:113]: the tests hardcode these values.  so this allowed us to change the inheritance hierarchy without rewriting *all* the tests.

File: src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+           codeActionTypeName + ";" +
+           options.ThroughMember?.Name;
+    }
+
+    private static async IAsyncEnumerable<ImplementInterfaceOptions> GetImplementOptionsAsync(
+        Document document, IImplementInterfaceInfo state, [EnumeratorCancellation] CancellationToken cancellationToken)
+    {
+        var compilation = await document.Project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+        if (state.MembersWithoutExplicitOrImplicitImplementationWhichCanBeImplicitlyImplemented.Length > 0)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs:132]: same logic that used to make all the code actions.  now it makes the 'options' at the service layer for what possible ways there are to implement. the code action will then pass this options object to the service layer to do the actual impl.

File: src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+            if (ShouldImplementDisposePattern(compilation, state, explicitly: true))
+                yield return new() { ImplementDisposePattern = true, Explicitly = true };
+        }
+
+        if (AnyImplementedImplicitly(state))
+            yield return new() { Explicitly = true };
+    }
+
+    private static bool AnyImplementedImplicitly(IImplementInterfaceInfo state)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs:176]: just a move.

File: src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+            {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private static ImmutableArray<ISymbol> GetDelegatableMembers(
+        Document document, IImplementInterfaceInfo state, CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs:202]: just a move.

File: src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceService.cs [hunk 10]
```diff
@@ -42,130 +34,27 @@ protected AbstractImplementInterfaceService()
...
     }
 
-    public ImmutableArray<CodeAction> GetCodeActions(Document document, ImplementTypeGenerationOptions options, SemanticModel model, SyntaxNode node, CancellationToken cancellationToken)
+    public async Task<IImplementInterfaceInfo?> AnalyzeAsync(Document document, SyntaxNode interfaceType, CancellationToken cancellationToken)
     {
-        var state = State.Generate(this, document, model, node, cancellationToken);
-        return GetActions(document, options, state, cancellationToken).ToImmutableArray();
-    }
-
-    private IEnumerable<CodeAction> GetActions(
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceService.cs:69]: all this moved into the actual CodeFixProvider

File: src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceService.cs [hunk 11]
```diff
@@ -189,4 +78,29 @@ protected static SyntaxTriviaList CreateCommentTrivia(SyntaxGenerator generator,
...
+    public async Task<Document> ImplementInterfaceAsync(
+        Document document,
+        IImplementInterfaceInfo info,
+        ImplementInterfaceOptions options,
+        ImplementTypeGenerationOptions generationOptions,
+        CancellationToken cancellationToken)
+    {
+        if (options.ImplementDisposePattern)
+        {
+            var generator = new ImplementInterfaceWithDisposePatternGenerator(
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceService.cs:91]: still deferring to the existing 'generator' types that do some work and have some virtual behavior.  i don't like them, but i didn't want to churn too much.

File: src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceService_DisposePattern.cs [hunk 12]
```diff
@@ -35,89 +37,16 @@ internal abstract partial class AbstractImplementInterfaceService
         parameterOptions: SymbolDisplayParameterOptions.IncludeName | SymbolDisplayParameterOptions.IncludeType,
         miscellaneousOptions: SymbolDisplayMiscellaneousOptions.UseSpecialTypes);
 
-    private static IMethodSymbol? TryGetIDisposableDispose(Compilation compilation)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceService_DisposePattern.cs:38]: moved to ImplementHelpers type.

File: src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceService_DisposePattern.cs [hunk 13]
```diff
@@ -35,89 +37,16 @@ internal abstract partial class AbstractImplementInterfaceService
...
-                {
-                    return disposeMethod;
-                }
-            }
-        }
-
-        return null;
-    }
-
-    private static bool ShouldImplementDisposePattern(State state, bool explicitly)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceService_DisposePattern.cs:61]: moved to ImplementHelpers type.

File: src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceService_DisposePattern.cs [hunk 14]
```diff
@@ -35,89 +37,16 @@ internal abstract partial class AbstractImplementInterfaceService
...
         Document document,
         ImplementTypeGenerationOptions options,
-        State state,
+        IImplementInterfaceInfo state,
         bool explicitly,
         bool abstractly,
-        ISymbol? throughMember) : ImplementInterfaceCodeAction(service, document, options, state, explicitly, abstractly, onlyRemaining: !explicitly, throughMember)
+        ISymbol? throughMember) : ImplementInterfaceGenerator(service, document, options, state, explicitly, abstractly, onlyRemaining: !explicitly, throughMember)
     {
-        public static ImplementInterfaceWithDisposePatternCodeAction CreateImplementWithDisposePatternCodeAction(
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceService_DisposePattern.cs:97]: got rid of all the factories.

File: src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceService_DisposePattern.cs [hunk 15]
```diff
@@ -271,9 +199,7 @@ public override string Title
         private IMethodSymbol CreateDisposeInterfaceMethod(
             Compilation compilation,
             Document document,
-            INamedTypeSymbol classType,
             IMethodSymbol disposeMethod,
-            IFieldSymbol disposedValueField,
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceService_DisposePattern.cs:276]: simply unused.

File: src/Features/Core/Portable/ImplementInterface/IImplementInterfaceService.cs [hunk 16]
```diff
@@ -5,14 +5,56 @@
...
     Task<Document> ImplementInterfaceAsync(Document document, ImplementTypeGenerationOptions options, SyntaxNode node, CancellationToken cancellationToken);
-    ImmutableArray<CodeAction> GetCodeActions(Document document, ImplementTypeGenerationOptions options, SemanticModel model, SyntaxNode node, CancellationToken cancellationToken);
+
+    Task<IImplementInterfaceInfo?> AnalyzeAsync(Document document, SyntaxNode interfaceType, CancellationToken cancellationToken);
+    Task<Document> ImplementInterfaceAsync(
+        Document document,
+        IImplementInterfaceInfo info,
+        ImplementInterfaceOptions options,
+        ImplementTypeGenerationOptions generationOptions,
+        CancellationToken cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ImplementInterface/IImplementInterfaceService.cs:59]: core part of the fix.  the service now has no knowledge of code actions.  it simply analyzes the data passed in and gives info about what it figured out.  the code action then says "based on thnat, here are hte fixes the user coudl make" and then based on that, it calls back into the service to actually make the change.

File: src/Features/Core/Portable/ImplementInterface/ImplementHelpers.cs [hunk 17]
```diff
@@ -114,4 +115,147 @@ bool InitializerReferencesParameter(SyntaxNode? initializer, IParameterSymbol pa
             return false;
         }
     }
+
+    public static bool IsLessAccessibleThan(ISymbol? first, INamedTypeSymbol second)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ImplementInterface/ImplementHelpers.cs:119]: just a move.  no change.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ImplementInterface/ImplementHelpers.cs:119]: same with the methods below

File: src/Features/Core/Portable/ImplementInterface/ImplementInterfaceGenerator.cs [hunk 18]
```diff
@@ -53,112 +53,9 @@ internal partial class ImplementInterfaceCodeAction : CodeAction
...
-            // in the same manner.  Note: 'implement through member' means implementing the same interface through
-            // an applicable member with the same name in the destination.
-            return explicitly.ToString() + ";" +
-               abstractly.ToString() + ";" +
-               onlyRemaining.ToString() + ":" +
-               typeName + ";" +
-               assemblyName + ";" +
-               codeActionTypeName + ";" +
-               throughMember?.Name;
-        }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/ImplementInterface/ImplementInterfaceGenerator.cs:157]: title+equivkey moved to the actual codeaction.

File: src/Features/VisualBasic/Portable/ImplementInterface/VisualBasicImplementInterfaceCodeFixProvider.vb [hunk 19]
```diff
@@ -25,49 +22,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.ImplementInterface
         Public Sub New()
         End Sub
 
-        Public NotOverridable Overrides ReadOnly Property FixableDiagnosticIds As ImmutableArray(Of String)
-            Get
-                Return ImmutableArray.Create(BC30149)
-            End Get
-        End Property
-
-        Public NotOverridable Overrides Function GetFixAllProvider() As FixAllProvider
```
Comment (CyrusNajmabadi) [inline src/Features/VisualBasic/Portable/ImplementInterface/VisualBasicImplementInterfaceCodeFixProvider.vb:34]: all the common code moved to base class.

File: eng/targets/Settings.props [hunk 20]
```diff
@@ -66,6 +66,9 @@
       Related to https://github.com/dotnet/sdk/issues/41791.
     -->
     <BuildWithNetFrameworkHostedCompiler>false</BuildWithNetFrameworkHostedCompiler>
+
+    <!-- https://github.com/dotnet/msbuild/issues/10306 -->
+    <CoreCompileDependsOn>$(CoreCompileDependsOn);ResolveKeySource</CoreCompileDependsOn>
```
Comment (tmat) [inline eng/targets/Settings.props:71]: Probably don't want to include this change in the PR

Comment (CyrusNajmabadi) [reply eng/targets/Settings.props:71]: ah true.  i will remove.  i need this locally @jjonescz  :)

Comment (CyrusNajmabadi) [inline eng/targets/Settings.props:71]:
```suggestion
```

File: src/Features/Core/Portable/ImplementInterface/ImplementInterfaceGenerator.cs [hunk 21]
```diff
@@ -28,42 +28,33 @@ private sealed partial class ImplementInterfaceGenerator
...
-        private readonly bool Abstractly;
-        private readonly bool OnlyRemaining;
-        private readonly bool ImplementDisposePattern;
+        private readonly AbstractImplementInterfaceService Service;
 
-        private readonly ISymbol? ThroughMember;
-        private readonly ImplementTypeGenerationOptions Options;
         private readonly IImplementInterfaceInfo State;
-        private readonly AbstractImplementInterfaceService Service;
+        private readonly ImplementTypeGenerationOptions GenerationOptions;
```
Comment (tmat) [inline src/Features/Core/Portable/ImplementInterface/ImplementInterfaceGenerator.cs:34]: In https://github.com/dotnet/roslyn/pull/74323 I removed ImplementTypeGenerationOptions, which grouped ImplementTypeOptions with other options that are no longer needed. So this should now be ImplementTypeOptions and you'd probably want to wait for that change and rename GenerationOptions.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ImplementInterface/ImplementInterfaceGenerator.cs:34]: yes.  will do that.

File: src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs [hunk 22]
```diff
@@ -59,14 +59,16 @@ public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
         }
     }
 
-    private IEnumerable<IImplementInterfaceGenerator> GetGenerators(
+    private async IEnumerable<IImplementInterfaceGenerator> GetGenerators(
         Document document, ImplementTypeGenerationOptions options, IImplementInterfaceInfo? state, CancellationToken cancellationToken)
     {
         if (state == null)
```
Comment (tmat) [inline src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs]: This is a bit weird - can we avoid calling the method if `state` is null?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ImplementInterface/AbstractImplementInterfaceCodeFixProvider.cs]: you're looking at old code i think.  it's not like that anymore.

File: src/Features/Core/Portable/ImplementInterface/ImplementInterfaceGenerator.cs (hunk 21 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/ImplementInterface/ImplementInterfaceGenerator.cs]: done.

**Comment by CyrusNajmabadi** [comment]
@tmat I've merged your change in.  ptal.

**Review by tmat — APPROVED** [review]

**Pull request merged**